### PR TITLE
Added N0 timing implementation framework

### PR DESCRIPTION
**Pull request by Charlie-Zheng**
Frame timings collected and implemented for:
Everyone except for Nilou in E

https://docs.google.com/spreadsheets/d/1DkFeG8fEkmM84XW2jtYuNpNxyE1W7-XXMj0V6Bn6wrA/edit?usp=sharing

### CODE REVIEW

**Comment by k0l11** [comment]
LGTM after a short look over it aside from a few comments. @unleashurgeek you might also want to check this out.

File: internal/characters/xingqiu/burst.go [hunk 1]
```diff
@@ -154,6 +182,40 @@ func (c *char) summonSwordWave() {
...
+	} else {
+		return XQ_N0_delays[activeCharKey]
+	}
+}
+
+func (c *char) burstStateDelayFunc() {
+	//ignore if on ICD
+	if c.StatusIsActive(burstICDKey) {
+		return
+	}
```
Comment (unleashurgeek) [inline internal/characters/xingqiu/burst.go:204]: should add a check for if the character is the same as when it was scheduled. ideally this check should instead be if the current NA is the same as the one that was scheduled, but I can't think of a good way to do that today. Should probably just add a TODO for this and create an issue

File: internal/characters/xingqiu/burst.go [hunk 2]
```diff
@@ -25,6 +29,30 @@ func init() {
...
+	XQ_N0_delays_alt_forms[keys.Raiden] = 13
+	alt_form_status_keys[keys.Raiden] = "raidenburst"
+
+	XQ_N0_delays[keys.Bennett] = 7
+	XQ_N0_delays[keys.Diluc] = 15
+	XQ_N0_delays[keys.Kazuha] = 10
+	XQ_N0_delays[keys.Keqing] = 8
+	XQ_N0_delays[keys.Xiangling] = 7
+
+	// jumping/dashing during the NA windup for some catalysts modifies their frames - said by koli
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/xingqiu/burst.go:96]: Hmm but holding the button for a certain duration definitely affects it. Kokomi spawns her N1 in like 2f, but you can hold for 4 frames then release and still do a N1.

File: internal/characters/xingqiu/burst.go [hunk 3]
```diff
@@ -154,6 +182,40 @@ func (c *char) summonSwordWave() {
...
+
+func (c *char) burstStateDelayFunc() {
+	//ignore if on ICD
+	if c.StatusIsActive(burstICDKey) {
+		return
+	}
+
+	state := c.Core.Player.CurrentState()
+	//this should start a new ticker if not on ICD and state is correct
+	if state == action.NormalAttackState {
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/xingqiu/burst.go]: I've used @unleashurgeek's suggestion of combining the conditions

File: internal/characters/xingqiu/burst.go (hunk 1 above)
Comment (Charlie-Zheng) [reply internal/characters/xingqiu/burst.go]: I added a source frame check, hopefully that ensure it works properly?

File: internal/characters/xingqiu/burst.go [hunk 4]
```diff
@@ -25,6 +29,30 @@ func init() {
...
+		XQ_N0_delays_alt_forms[i] = -1
+	}
+
+	XQ_N0_delays[keys.Nahida] = 9
+	XQ_N0_delays[keys.Xingqiu] = 7
+	XQ_N0_delays[keys.Yelan] = 9
+
+	XQ_N0_delays[keys.Raiden] = 13
+	XQ_N0_delays_alt_forms[keys.Raiden] = 13
+	alt_form_status_keys[keys.Raiden] = "raidenburst"
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/xingqiu/burst.go]: I've made the burst key and melee key for Raiden and Tartaglia, respectively, public.

File: internal/characters/xingqiu/burst.go [hunk 5]
```diff
@@ -25,6 +31,80 @@ func init() {
...
+		XQ_N0_delays_alt_forms[i] = -1
+	}
+
+	XQ_N0_delays[keys.Nahida] = 9
+	XQ_N0_delays[keys.Xingqiu] = 7
+	XQ_N0_delays[keys.Yelan] = 9
+
+	XQ_N0_delays[keys.Raiden] = 13
+	XQ_N0_delays_alt_forms[keys.Raiden] = 13
+	alt_form_status_keys[keys.Raiden] = raiden.BurstKey
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/xingqiu/burst.go:45]: It's not, I can remove it. Previous this was the only alt form I had counted so I had it included to show the difference.

File: internal/data/NAstateframes.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+import (
+	"github.com/genshinsim/gcsim/internal/characters/ayato"
+	"github.com/genshinsim/gcsim/internal/characters/cyno"
+	"github.com/genshinsim/gcsim/internal/characters/tartaglia"
+	"github.com/genshinsim/gcsim/pkg/core/keys"
+)
+
+var PercentDelay5 = make([]int, keys.EndCharKeys)
+var PercentDelay5AltForms = make([]int, keys.EndCharKeys)
+var AltFormStatusKeys = make([]string, keys.EndCharKeys)
```
Comment (unleashurgeek) [inline internal/data/NAstateframes.go:12]: NIT: improve naming

also change file name to all lowercase name

File: internal/characters/thoma/burst.go [hunk 7]
```diff
@@ -78,56 +79,100 @@ func (c *char) Burst(p map[string]int) action.ActionInfo {
...
+			Write("state", action.NormalAttackState).
+			Write("icd", c.StatusExpiry(burstICDKey))
+		c.burstTickSrc = c.Core.F
+		//use the hitlag affected queue for this
+		c.QueueCharTask(c.burstTickerFunc(c.Core.F), 60) //check every 1sec
+	}
+}
+
+func (c *char) burstStateHook() {
 	c.Core.Events.Subscribe(event.OnStateChange, func(args ...interface{}) bool {
```
Comment (k0l11) [inline internal/characters/thoma/burst.go:115]: This hook can be changed to OnAttack. Then you can skip the animation state check in the subscription here. Same applies to Xingqiu.

OnAttack gets emitted here:
https://github.com/genshinsim/gcsim/blob/97e8e28c0e21004127ac33dddc62faa91331fff7/pkg/core/player/exec.go#L157
OnStateChange gets emitted here:
https://github.com/genshinsim/gcsim/blob/c184fb9382670f65c852e6a568bd51819fe0fe65/pkg/core/player/animation/animation.go#L87

As you can see, OnAttack gets emitted right before SetActionUsed gets called, which is where OnStateChange is emitted. You don't need info about the previous state/action, so OnAttack should also work and reduce the amount times this function gets called.

File: internal/common/normalattackstate.go [hunk 8]
```diff
@@ -134,3 +140,146 @@ func Get5PercentN0Delay(activeChar *character.CharWrapper) int {
...
+		abilICDKey:  abilICDKey,
+		abilTickSrc: 0,
+		abilHookSrc: 0,
+		delayFunc:   delayFunc,
+		summonFunc:  summonFunc,
+	}
+
+}
+
+func (w *NAHook) NAStateHook() {
```
Comment (unleashurgeek) [inline internal/common/normalattackstate.go:178]: NIT: I'd name this something like `Attach` or `Enable` to make it clear that it is performing a subscription

File: internal/common/normalattackstate.go [hunk 9]
```diff
@@ -134,3 +140,146 @@ func Get5PercentN0Delay(activeChar *character.CharWrapper) int {
...
+		w.abilHookSrc = w.Core.F
+		delay := w.delayFunc(w.Core.Player.ActiveChar())
+		w.Core.Log.NewEvent(fmt.Sprintf("%v delay on state change", w.abilName), glog.LogCharacterEvent, w.c.Index).
+			Write("delay", delay)
+		// This accounts for the delay in n0 timing needed to trigger
+
+		if delay > 0 {
+
+			w.Core.Tasks.Add(w.naStateDelayFuncGen(w.Core.F), delay)
+		} else {
```
Comment (unleashurgeek) [inline internal/common/normalattackstate.go:193]: return inside this `if` and remove `else` to reduce code depth

NIT: remove blank line

File: internal/common/normalattackstate.go [hunk 10]
```diff
@@ -134,3 +140,146 @@ func Get5PercentN0Delay(activeChar *character.CharWrapper) int {
...
+			} else {
+				w.summonFunc()
+				w.c.AddStatus(w.abilICDKey, 60, true)
+				w.Core.Log.NewEvent(fmt.Sprintf("%v on state change", w.abilName), glog.LogCharacterEvent, w.c.Index).
+					Write("state", action.NormalAttackState).
+					Write("icd", w.c.StatusExpiry(w.abilICDKey))
+				w.abilTickSrc = w.Core.F
+				//use the hitlag affected queue for this
+				w.c.QueueCharTask(w.naTickerFunc(w.Core.F), w.abilProcICD) //check every `ICD`` frames
+			}
```
Comment (unleashurgeek) [inline internal/common/normalattackstate.go:210]: IMO I think it'd be cleaner if we flip this condition and then have a single log line. Something like:

```go
isActive := w.c.StatusIsActive(w.abilICDKey)
if !isActive {
  w.summonFunc()
  w.c.AddStatus(w.abilICDKey, 60, true)
  w.c.QueueCharTask(w.naTickerFunc(w.Core.F), w.abilProcICD)
  w.abilTickSrc = w.Core.F
}

w.Core.Log.NewEvent(
    fmt.Sprintf("%v on state change [triggered=%v]", w.abilName, !isActive),
    glog.LogCharacterEvent,
    w.c.Index)
  .Write("state", w.Core.Player.CurrentState())
  .Write("icd", w.c.StatusExpiry(w.abilICDKey))
```

File: internal/common/normalattackstate.go [hunk 11]
```diff
@@ -134,3 +140,146 @@ func Get5PercentN0Delay(activeChar *character.CharWrapper) int {
...
+		w.summonFunc()
+		w.c.AddStatus(w.abilICDKey, 60, true)
+		w.Core.Log.NewEvent(fmt.Sprintf("%v on state change", w.abilName), glog.LogCharacterEvent, w.c.Index).
+			Write("state", action.NormalAttackState).
+			Write("icd", w.c.StatusExpiry(w.abilICDKey))
+		w.abilTickSrc = w.Core.F
+		//use the hitlag affected queue for this
+		w.c.QueueCharTask(w.naTickerFunc(w.Core.F), w.abilProcICD) //check every `ICD` frames
+	}
+}
```
Comment (unleashurgeek) [inline internal/common/normalattackstate.go:240]: This logic seems very similar/nearly identical to the `delay = 0` case in the subscription? can we reuse it in some way (have the different delayed check -> function that does status check + all the other necessary calls?)

File: internal/common/normalattackstate.go [hunk 12]
```diff
@@ -134,3 +140,146 @@ func Get5PercentN0Delay(activeChar *character.CharWrapper) int {
...
+			Write("state", state).
+			Write("icd", w.c.StatusExpiry(w.abilICDKey))
+
+		//we can trigger here b/c we're in normal state still and src is still the same
+		w.summonFunc()
+		w.c.AddStatus(w.abilICDKey, 60, true)
+		//in theory this should not hit an icd?
+		//use the hitlag affected queue for this
+		w.abilTickSrc = w.Core.F
+		w.c.QueueCharTask(w.naTickerFunc(w.Core.F), w.abilProcICD) //check every 1sec
```
Comment (unleashurgeek) [inline internal/common/normalattackstate.go:283]: Same comment with the other two places. Seems like this log + 4 operations can be reused

File: internal/characters/xingqiu/xingqiu.go [hunk 13]
```diff
@@ -38,7 +37,15 @@ func NewChar(s *core.Core, w *character.CharWrapper, _ profile.CharacterProfile)
...
+	common.NewNAHook(
+		c.CharWrapper,
+		c.Core,
+		"xingqiu burst",
+		burstKey,
+		60,
+		burstICDKey,
+		common.Get5PercentN0Delay,
+		c.summonSwordWave).
+		NAStateHook()
```
Comment (unleashurgeek) [inline internal/characters/xingqiu/xingqiu.go:49]: super nit: does `gofmt` do this? IMO `.` should always start the next line when you are splitting between multiple (but also go syntax weird so whatever lol)

File: internal/characters/thoma/thoma.go [hunk 14]
```diff
@@ -42,6 +42,17 @@ func (c *char) Init() error {
 		c.c6buff = make([]float64, attributes.EndStatType)
 		c.c6buff[attributes.DmgP] = .15
 	}
+
+	common.NewNAHook(
```
Comment (k0l11) [inline internal/characters/thoma/thoma.go:46]: Please consider passing a `common.NAHook` struct directly to make it easier to read/harder to mix up the params.

**Comment by k0l11** [comment]
closes #1151 
need to check whether #989 is still a problem

File: internal/common/normalattackstate.go [hunk 15]
```diff
@@ -147,139 +147,128 @@ func Get0PercentN0Delay(activeChar *character.CharWrapper) int {
...
+	icd := w.C.StatusIsActive(w.AbilICDKey)
+	if icd {
+		w.Core.Log.NewEvent(fmt.Sprintf("%v not triggered, on proc ICD", w.AbilName), glog.LogCharacterEvent, w.C.Index).
+			Write("icd", w.C.StatusExpiry(w.AbilICDKey))
+	}
+	return !icd
+}
+
+func (w *NAHook) checkState() bool {
+	state := w.Core.Player.CurrentState()
```
Comment (unleashurgeek) [inline internal/common/normalattackstate.go:249]: NIT: use `state` variable in logs or remove

**Review by unleashurgeek — APPROVED** [review]

**Pull request merged**